Global error handler cleanup - Trace SDK

lalitb opened the pull request:
Fixes #
Design discussion issue (if applicable) #

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -129,9 +129,9 @@ impl Default for Config {
-                            Error::Other(String::from(
-                                "Missing or invalid OTEL_TRACES_SAMPLER_ARG value. Falling back to default: 1.0"))
+                        otel_warn!(
+                            name: "TracesSampler.TraceIdRatio.Error",
```
the name make it sound like there is something wrong with TraceSampler, but the issue is with invalid configuration.

lalitb replied:
done

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -129,9 +129,9 @@ impl Default for Config {
-                                "Missing or invalid OTEL_TRACES_SAMPLER_ARG value. Falling back to default: 1.0"))
+                        otel_warn!(
+                            name: "TracesSampler.TraceIdRatio.Error",
+                            message= "Missing or invalid OTEL_TRACES_SAMPLER_ARG env variable. Falling back to default: 1.0"
```
need more user actionable message.
OTEL_TRACES_SAMPLER env variable has the value traceidratio and hence ratio should be provided by OTEL_TRACES_SAMPLER_ARG env variable which is missing/invalid. TraceIdRatioBased sampler will still be enabled, but will a fallback value of 1.0 (i.e 100% traces are accepted) for ratio.

lalitb replied:
done

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -129,9 +129,9 @@ impl Default for Config {
-                            Error::Other(String::from(
-                                "Missing or invalid OTEL_TRACES_SAMPLER_ARG value. Falling back to default: 1.0"))
+                        otel_warn!(
+                            name: "Sampler.TraceIdRatio.InvalidArgument",
```
the name is misleading. This is not an issue occurring in the Sampler (it affects sampler), but rather a configuration issue.

lalitb replied:
I didn't get this - isn't this the configuration for sampler which is causing the issue? In that case, it is related to sampler.

cijothomas replied:
this is purely configuration issue, that *affects* sampler, not an issue inside sampler. Sampler's code is not even executed, so the name is misleading, as it seem to indicate some issue inside Sampler. If this log is emitted from inside a Sampler's `new()` method, then the name makes sense, but in this case, it is occurring outside Sampler.

lalitb replied:
updated as below:
```rust
    name: "TracerProvider.Config.Sampler.InvalidArgument",
    message = "OTEL_TRACES_SAMPLER is set to 'traceidratio' but OTEL_TRACES_SAMPLER_ARG environment variable is missing or invalid. OTEL_TRACES_SAMPLER_ARG must be a valid float between 0.0 and 1.0 representing the desired sampling probability (0.0 = no traces sampled, 1.0 = all traces sampled, 0.5 = 50% of traces sampled). Falling back to default ratio: 1.0 (100% sampling)"
 );
```

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -149,37 +149,45 @@ impl Default for Config {
-                    )));
+                    otel_warn!(
+                        name: "Sampler.ParentBasedJaegerRemote.Unsupported",
+                        message = "ParentBased JaegerRemote sampler is not implemented in this SDK version. Using fallback sampler: ParentBased(AlwaysOn). Configure an alternative sampler using OTEL_TRACES_SAMPLER"
```
OTEL_TRACES_SAMPLER env variable has the value jaeger-remote, which is unsupported. Will fallback to using ParentBased(AlwaysOn) sampler.

lalitb replied:
rephrased this here, and also other places.

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -134,7 +134,10 @@ impl SpanProcessor for SimpleSpanProcessor {
-            global::handle_error(err);
+            otel_error!(
+                name: "SimpleProcessor.OnEnd.Error",
+                reason = format!("{:?}", err)
```
do we need error severity? would the information here be duplicate of what an exporter might already be logging?
I am also unsure if the message is really actionable by end users. there is no indication of what happened or what is the impact of this..

lalitb replied:
> would the information here be duplicate of what an exporter might already be logging?

Not necessarily - exporter may choose to return the export status to be logged here. The status may be actionable in that case. For logs, we check the error type and then decide whether to log as `error` or `debug`. Similar logic needs to be done here. Will add a TODO here. Meanwhile, do you want to change this to `debug` ?

lalitb replied:
Added TODO, and changed to otel_debug.

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -354,7 +367,10 @@ impl<R: RuntimeChannel> BatchSpanProcessorInternal<R> {
                         if let Err(err) = export_task.await {
-                            global::handle_error(err);
+                            otel_error!(
+                                name: "BatchSpanProcessor.Export.Error",
```
There is another log with the name "BatchProcessor.OnEnd.Error". Is this duplicate ?

lalitb replied:
Good catch. We have `SimpleProcessor.OnEnd.Error`. To make it consistent, this can be modified to `BatchProcessor.OnEnd.Error`

lalitb replied:
These are the event types:

`SimpleProcessor.OnEnd.Error`: export error
`BatchSpanProcessor.OnEnd.Error`: export error
`BatchSpanProcessor.OnEnd.SendResultError`: error sending export status/

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -313,14 +319,21 @@ impl<R: RuntimeChannel> BatchSpanProcessorInternal<R> {
+                // If no channel is provided and the export operation encountered an error,
+                // log the error directly here.
+                otel_error!(
+                    name: "BatchSpanProcessor.Flush.ExportError",
```
unsure why we need Error level internal log for Flush operation? Aren't we returning Flush `Result` back to user already?

lalitb replied:
It doesn't in existing code, can add a TODO. For tracing, the PR is doing search-replace for the global handler.

lalitb replied:
added TODO.

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -246,7 +249,10 @@ impl<R: RuntimeChannel> SpanProcessor for BatchSpanProcessor<R> {
 
         if let Err(err) = result {
-            global::handle_error(TraceError::Other(err.into()));
+            otel_error!(
```
will this be triggered when channel is full? if yes, we shouldn't do such logging as that can overwhelm internal logs

lalitb replied:
yes it will. Will make it debug in that case.

lalitb replied:
updated as below:

```rust
        let result = self.message_sender.try_send(BatchMessage::ExportSpan(span));

        if let Err(err) = result {
            otel_debug!(
                name: "BatchSpanProcessor.OnEnd.SendResultError",
                reason = format!("{:?}", TraceError::Other(err.into()))
            );
        }
```

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -149,37 +149,45 @@ impl Default for Config {
-                        Error::Other(format!("Unrecognised OTEL_TRACES_SAMPLER value: {}. Falling back to default: parentbased_always_on",
-                        s
-                    )));
+                    otel_warn!(
```
name: TracerProvider.Configuration.InvalidSampler
message : the why which includes the list of supported values, and also the value input by user. the impact - usually fallback part.

some thoughts above.

lalitb replied:
upated as below:
```rust
otel_warn!(
    name: "TracerProvider.Config.Sampler.InvalidSamplerType",
    message = format!(
         "Unrecognized sampler type '{}' in OTEL_TRACES_SAMPLER environment variable. Valid values are: always_on, always_off, traceidratio, parentbased_always_on, parentbased_always_off, parentbased_traceidratio. Using fallback sampler: ParentBased(AlwaysOn)", s
     ),
     sampler = s
);
```

cijothomas replied:
TracerProvider.Config.Sampler.InvalidSamplerType can be shortened, as it repeats Sampler twice.

ThomsonTan commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -129,9 +129,9 @@ impl Default for Config {
-                                "Missing or invalid OTEL_TRACES_SAMPLER_ARG value. Falling back to default: 1.0"))
+                        otel_warn!(
+                            name: "TracerProvider.Config.Sampler.InvalidArgument",
+                            message = "OTEL_TRACES_SAMPLER is set to 'traceidratio' but OTEL_TRACES_SAMPLER_ARG environment variable is missing or invalid. OTEL_TRACES_SAMPLER_ARG must be a valid float between 0.0 and 1.0 representing the desired sampling probability (0.0 = no traces sampled, 1.0 = all traces sampled, 0.5 = 50% of traces sampled). Falling back to default ratio: 1.0 (100% sampling ...
```
Include the current (invalid) value of `OTEL_TRACES_SAMPLER_ARG` in the warning message?

lalitb replied:
Done:
```
otel_warn!(
    name: "TracerProvider.Config.Sampler.InvalidArgument",
    message = "OTEL_TRACES_SAMPLER is set to 'traceidratio' but OTEL_TRACES_SAMPLER_ARG environment variable is missing or invalid. OTEL_TRACES_SAMPLER_ARG must be a valid float between 0.0 and 1.0 representing the desired sampling probability (0.0 = no traces sampled, 1.0 = all traces sampled, 0.5 = 50% of traces sampled). Falling back to default ratio: 1.0 (100% sampling)",
    otel_traces_sampler_arg = format!("{:?}", sampler_arg)
);
```

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -354,7 +369,10 @@ impl<R: RuntimeChannel> BatchSpanProcessorInternal<R> {
                         if let Err(err) = export_task.await {
-                            global::handle_error(err);
+                            otel_error!(
+                                name: "BatchSpanProcessor.OnEnd.Error",
```
this doesn't look correct to me, as this is occurring in background thread/task, and *not* in BSP's OnEnd method.

lalitb replied:
Good point. Changed it to `BatchSpanProcessor.Export.Error`, to be consistent with `BatchLogProcessor.Export.Error`, as the error occurs while processing export message.

lalitb commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -143,43 +144,52 @@ impl Default for Config {
+                            "Unrecognized sampler type '{}' in OTEL_TRACES_SAMPLER environment variable. Valid values are: always_on, always_off, traceidratio, parentbased_always_on, parentbased_always_off, parentbased_traceidratio. Using fallback sampler: ParentBased(AlwaysOn)",
+                            s
+                        ),
+                        sampler = s
```
Just thought separate attributes are in general helpful as search option in backend queries, mayn't be in this case. Will remove it as attribute, as message looks incomplete without it.

lalitb replied:
this is done./

cijothomas commented on opentelemetry-sdk/src/trace/config.rs:
```diff
@@ -143,43 +144,51 @@ impl Default for Config {
-                        s
-                    )));
+                    otel_warn!(
+                        name: "TracerProvider.Config.Sampler.InvalidSamplerType",
```
TracerProvider.Config.Sampler.InvalidSamplerType
TracerProvider.Config.Sampler.Unsupported
TracerProvider.Config.Sampler.InvalidArgument

Lets use consistent naming. The 1st one has Sampler word repeated, while others don't.
A shorter name like `TracerProvider.InvalidSamplerType` should be sufficient IMHO.
Alternatively, name can be `TracerProviderInvalidConfiguration`, and message can explain what went wrong, and additional attributes can be used to describe the component being configured incorrectly.

lalitb replied:
Have named it to:
TracerProvider.Config.InvalidSamplerArgument
TracerProvider.Config.UnsupportedSampler
TracerProvider.Config.InvalidSamplerType

Wanted to keep 3 levels of naming as done in other places. Let me know if this still doesn't look consistent.

cijothomas replied:
From consistency, this is better. Though I am unsure if the name is easily followable by end users. User may not have ever touched/used `Config` at all, and when they see 'Config' in event name, it may not be obvious. Good to follow up later.

cijothomas commented on opentelemetry-sdk/src/trace/span_processor.rs:
```diff
@@ -246,7 +250,10 @@ impl<R: RuntimeChannel> SpanProcessor for BatchSpanProcessor<R> {
         if let Err(err) = result {
-            global::handle_error(TraceError::Other(err.into()));
+            otel_debug!(
+                name: "BatchSpanProcessor.Export.SendResultError",
```
this is confusing. This is the `on_end` method.. Given this is BatchProcessor, this will only queue up spans on `on_end` and export occurs asynchronously. The name "BatchSpanProcessor.Export.SendResultError" seem to indicate that some sort of Export is going on here, which is incorrect/misleading, as export never occurs on `on_end` call for BatchSpanProcessor. Also "SendResultError" is not reflecting anything going on here...

lalitb replied:
Good catch. Have named it to `BatchSpanProcessor.OnEnd.ExportQueueingFailed`.

cijothomas reviewed (approved):
Good to merge now,
Have to revisit all the logs here in the future to ensure we have the right balance between too much vs too little.

cijothomas commented:
@lalitb There are quite large number of unresolved conversations in the PR, hard to tell if they are still relevant. Could you resolve the ones that are addressed and merge?

lalitb commented:
> @lalitb There are quite large number of unresolved conversations in the PR, hard to tell if they are still relevant. Could you resolve the ones that are addressed and merge?

I have addressed all the comments, so resolved them. Merging now. Thanks for the thorough review
